runtime/src/main/java/org/corfudb/runtime/view/stream/CompositeStreamAddressSpace.java

```diff
@@ -0,0 +1,338 @@
+package org.corfudb.runtime.view.stream;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.UUID;
+import java.util.function.Function;
+
+import org.corfudb.protocols.wireprotocol.ILogData;
+import org.corfudb.protocols.wireprotocol.Token;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.exceptions.TrimmedException;
+import org.corfudb.runtime.view.Address;
+import org.corfudb.runtime.view.StreamOptions;
+
+/**
+ * This class defines the space of addresses of a stream under the assumption that a stream can
+ * be checkpoint(ed). It is made up of two separate spaces: that of regular addresses
+ * and that of the checkpoint addresses. The CompositeStreamAddressSpace handles both spaces
+ * to transparently provide a single resolved view of the stream address space to upper layers.
+ *
+ * Created by amartinezman on 4/24/18.
+ */
+public class CompositeStreamAddressSpace implements IStreamAddressSpace {
+
+    private BackpointerStreamAddressSpace regularSAS;
+    private CheckpointStreamAddressSpace cpSAS;
+    private boolean pointerOnRegularStream;
+    final private CorfuRuntime r;
+    final private UUID regularStreamId;
+    final private UUID cpStreamId;
+
+    public CompositeStreamAddressSpace(UUID id, CorfuRuntime runtime) {
+        regularStreamId = id;
+        cpStreamId = CorfuRuntime.getCheckpointStreamIdFromId(id);
+        r = runtime;
+        regularSAS = new BackpointerStreamAddressSpace(regularStreamId, runtime);
+        cpSAS = new CheckpointStreamAddressSpace(cpStreamId, runtime);
+        pointerOnRegularStream = true;
+    }
+
+    @Override
+    public void setStreamOptions(StreamOptions options) {
+        this.regularSAS.setStreamOptions(options);
+        this.cpSAS.setStreamOptions(options);
+    }
+
+    @Override
+    public void reset() {
+        regularSAS.reset();
+        cpSAS.reset();
+        pointerOnRegularStream = true;
+    }
+
+    @Override
+    public void seek(long address) {
+        // Seek for the address on the regular stream, if it is not available it means
+        // it has been checkpoint(ed) and trimmed.
```

Comment:
I think this changes the behavior of seek. Currently, on master, when you seek to an address that doesn't belong to the stream, the stream layer will choose the "closest" (i.e. greatest stream address less than the seek address) address to the non-stream address and not throw an Exception.

There's code in mp that queries the global tail and does stream.seek(globalTail) in that case if the stream doesn't have the global tail then it will break.